pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+	fn do_init_collator(who: &T::AccountId) -> DispatchResult {
+		T::Currency::mint_into(
+			T::CollatorCurrencyId::get(),
+			who,
+			T::DefaultCollatorStake::get(),
+		)?;
+		T::Rewards::deposit_stake(
+			(T::Domain::get(), T::CollatorCurrencyId::get()),
+			who,
+			T::DefaultCollatorStake::get(),
+		)
+	}
+
+	/// Withdraw currently staked amount for target address and immediately burn it.
+	/// Disables receiving rewards onwards.
+	fn do_exit_collator(who: &T::AccountId) -> DispatchResult {
+		let amount =
+			T::Rewards::account_stake((T::Domain::get(), T::CollatorCurrencyId::get()), who);
+		T::Rewards::withdraw_stake(
+			(T::Domain::get(), T::CollatorCurrencyId::get()),
+			who,
+			amount,
+		)?;
+		T::Currency::burn_from(T::CollatorCurrencyId::get(), who, amount).map(|_| ())
+	}
+
+	/// Mint remaining rewards if not everything was distributed to the groups.
+	///
+	/// NOTE: Noop if [RemainingRewardCollector] is `None`.
+	fn do_mint_remainder(
+		total_reward: T::Balance,
+		reward_results: &Vec<Result<T::Balance, DispatchError>>,
+	) -> DispatchResult {
+		if let Some(pallet_id) = T::RemainingRewardCollector::get() {
+			let remaining_rewards = reward_results.iter().fold(total_reward, |acc, result| {
+				// Don't subtract from total rewards if group reward failed
+				acc.saturating_sub(result.unwrap_or_default())
+			});
+			T::Currency::mint_into(
+				T::RewardCurrency::get(),
+				&pallet_id.into_account_truncating(),
+				remaining_rewards,
+			)?;
+		}
+
+		Ok(())
+	}
+
+	/// Apply epoch changes and distribute rewards.
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_epoch() -> Weight {
+		let mut groups = 0;
+		let mut weight_changes = 0;
+		let mut num_joining = 0;
+		let mut num_leaving = 0;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextEpochChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveEpochData::<T>::try_mutate(|epoch_data| {
+					num_joining = changes.collators.inc.len();
+					num_leaving = changes.collators.out.len();
+					for leaving in changes.collators.out.drain(..) {
+						Self::do_exit_collator(&leaving)?;
+					}
+					for joining in changes.collators.inc.drain(..) {
+						Self::do_init_collator(&joining)?;
+					}
+
+					groups = T::Rewards::distribute_reward_with_weights(
+						epoch_data.reward,
+						epoch_data.weights.iter().map(|(g, w)| (*g, *w)),
+					)
+					.map(|results| {
+						Self::do_mint_remainder(epoch_data.reward, &results)?;
+						Ok(results.len() as u32)
+					})
+					.and_then(|num_groups| num_groups)?;
+
+					for (&group_id, &weight) in &changes.weights {
+						epoch_data.weights.try_insert(group_id, weight).ok();
+						weight_changes += 1;
+					}
+
+					epoch_data.reward = changes.reward.unwrap_or(epoch_data.reward);
+
+					Self::deposit_event(Event::NewEpoch {
+						reward: epoch_data.reward,
+						last_changes: mem::take(changes),
+					});
+
+					Ok(())
+				})
+			})
+		})
+		.ok();
```

Comment:
This applies also to `pallet-liquidity-rewards`. Maybe we should add a `log::error()` or similar in case this doesn't work as expected to realize that. Wdyt? 
BTW, I do not how the log works on Substrate, is this something I could see from the `polkadot.js` or only in the terminal where I run the node?

Reply:
> BTW, I do not how the log works on Substrate, is this something I could see from the `polkadot.js` or only in the terminal where I run the node?

AFAIK, you can only see this in the node logs unfortunately.